### ISSUE TITLE
Unable to push image after upgrade to 1.12

### ISSUE THREAD

**Issue by zgoda** [docker/docker#25213]
**Output of `docker version`:**

```
$ docker version
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:11:10 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 22:11:10 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
$ docker info
Containers: 14
 Running: 0
 Paused: 0
 Stopped: 14
Images: 719
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 879
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.45 GiB
Name: JAZG-LENOVO
ID: 73QC:44RO:E4BA:YLI5:NAQT:OHDX:64YM:VJKF:5VFT:QLTS:VVMI:OORS
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Standard Ubuntu 16.04 desktop

**Steps to reproduce the issue:**
1. Build image with 1.10 and push to registry
2. Build the same image with 1.12
3. Push this image to registry

100% reproductible on my setup

**Describe the results you received:**

```
$ docker push myregistry.com/myimage:1.3.7
The push refers to a repository [myregistry.com/myimage]
reference does not exist
```

**Describe the results you expected:**

Push succeeds.

**Additional information you deem important (e.g. issue happens only occasionally):**

`myregistry.com` has properly trusted server certificate, client certificate is valid, client key and certificate are both in `/etc/docker/certs/myregistry.com`

Output of systemctl status docker.service:

```
● docker.service - Docker Application Container Engine
   Loaded: loaded (/lib/systemd/system/docker.service; enabled; vendor preset: enabled)
   Active: active (running) since pią 2016-07-29 11:22:08 CEST; 1min 12s ago
     Docs: https://docs.docker.com
 Main PID: 28184 (dockerd)
    Tasks: 18
   Memory: 15.4M
      CPU: 680ms
   CGroup: /system.slice/docker.service
           ├─28184 /usr/bin/dockerd -H fd://
           └─28191 docker-containerd -l unix:///var/run/docker/libcontainerd/docker-containerd.sock --shim docker-containerd-shim --metrics-interval=0 --start-timeout 2m --state-dir /var/run/docker/libcontainerd/containerd --runtime docker-runc

lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969141862+02:00" level=error msg="migration failed for b0ffef6dbd7de999cb61c5e0464329d416a96764d291abeb5948170ebb9ac44e, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969151339+02:00" level=error msg="migration failed for 131b8edd23fc40b5bb740f2c4e27eeaeb62331e32fe2c41b59cd69c39546ff9b, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969163942+02:00" level=error msg="migration failed for 6d5fe13ac3f888ad0e678580eee80372ac5bdff6c6536e3da961665400ead7e1, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969175774+02:00" level=error msg="migration failed for f48b5217e2a954ac6bbf58f41510d26bfbf1c92d7d33aa3b8a45bcd6b1a0ce4e, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969460214+02:00" level=error msg="migration failed for b0ffef6dbd7de999cb61c5e0464329d416a96764d291abeb5948170ebb9ac44e, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969470166+02:00" level=error msg="migration failed for 131b8edd23fc40b5bb740f2c4e27eeaeb62331e32fe2c41b59cd69c39546ff9b, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969477107+02:00" level=error msg="migration failed for 7a210d52e9205905407a55852f1ca39b8fcd259e45faa6a19fd606c7a5b492ef, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969484316+02:00" level=error msg="migration failed for f551bcd5a819b1b1d01c2423516fa3eda7c571d150747ca9460a50463139ee95, err: invalid checksum digest format"
lip 29 11:22:06 JAZG-LENOVO dockerd[28184]: time="2016-07-29T11:22:06.969739296+02:00" level=error msg="migration failed for b0ffef6dbd7de999cb61c5e0464329d416a96764d291abeb5948170ebb9ac44e, err: invalid checksum digest format"
lip 29 11:22:08 JAZG-LENOVO systemd[1]: Started Docker Application Container Engine.
```

**Comment by zgoda** [comment]
Downgrade to 1.11 or 1.10 does not help, image is not being pushed and the log shows the same kind of errors.

**Comment by zgoda** [comment]
More from journalctl:

```
lip 29 12:01:30 JAZG-LENOVO dockerd[3005]: time="2016-07-29T12:01:30.015663763+02:00" level=error msg="Attempting next endpoint for push after error: tag does not exist: myregistry.com/myimage:1.3.7"
lip 29 12:01:31 JAZG-LENOVO dockerd[3005]: time="2016-07-29T12:01:31.679076989+02:00" level=error msg="Not continuing with push after error: reference does not exist"
```

**Comment by zgoda** [comment]
Registry is running and is accessible:

```
$ curl --cert client.cert.pem --key client.key.pem -X GET https://myregistry.com/v1/repositories/myimage/tags
{"1.3.2": "a71e48e89659c493139f5b17393c3d7f68e0c9e66ecb7f0ef68df3f3b4765730", "1.3.3": "3cfc9ed9736c81314f15569f2c200bbf796b51e431af9993509999633e2a0aa8", "1.3.4": "e5471a0df31d48e7ded569e65cd0ab64665dbf2fc1abd67409bf0df7e77b36cd", "1.3.5": "25ea179084eef6e95b6ba22e75aa203de0e0ef63ca45fd1b46fb224e44c3adad", "1.3.6": "8a4add655fea1a4b90204d9b5eae4d9a2894bcf985f8709144c8ee4042793e11", "latest": "8a4add655fea1a4b90204d9b5eae4d9a2894bcf985f8709144c8ee4042793e11"}
```

(registry is behind HAProxy that requires client SSL authentication)

**Comment by thaJeztah** [comment]
ping @aaronlehmann @dmcgowan PTAL

**Comment by dmcgowan** [comment]
Looks like migration is failing `msg="migration failed for b0ffef6dbd7de999cb61c5e0464329d416a96764d291abeb5948170ebb9ac44e, err: invalid checksum digest format"`. Did you recently update your engine from pre-1.10 as well? When running 1.12 and issuing `docker images`, do you see all the images you expect? Can you also try re-pulling the image you are using as the base of Dockerfile (the `From` image).

See more of the docker engine log may also be helpful as well, since it might indicate which images those failed layers belonged to.

**Comment by aaronlehmann** [comment]
This looks like #20267, a bug in migration where temporary files were written non-atomically, so they could be corrupted if the process was interrupted. It was fixed in 1.10.2. If 1.10.0 or 1.10.1 was installed at some point and was interrupted during the migration process, it could have caused this corruption.

The fix would be to delete the `.migration-diffid` files in the subdirectories of `/var/lib/docker/graph`.

**Comment by zgoda** [comment]
I'm very sorry, this is classic PEBKAC.

In one specific case tagging line has been missing from build script.
